### DISCUSSION TITLE
Structuring Jotai for state switches when a component unmounts/changes data

### DISCUSSION THREAD

**Discussion by okbytes** [pmndrs/jotai#1182] [Q&A]
Using Jotai for global state has been a delight. 

However, I'm struggling on how to structure code for "global to a specific part of the component tree" that will clear its state when the component unmounts or the data changes. Here is some pseudo-code outlining the structure today using React.Context:

```jsx
function ConfigComponent(props) {
    const [state, dispatch] = useReducer(reducer, initialState);
    useQuery(GetDataDoc, {
        variables: { id: props.id },
        onCompleted: data => dispatch({ type: "replace", data })
    })

    //... state is used in the body of the function, too

    return (
        <ReactContextProvider state={state} dispatch={dispatch}>
            {/*
                ...lots of sub-components
            */}
        </ReactContextProvider>
    )
}
```

I'm using this approach to help pass down the state and reducer's dispatch function to the many sub-components. It helps, too, that when ConfigComponent gets remounted, it will fetch the id it's given and push down all the relevant data to the sub-components.

Jotai should be able to provide better render performance and ease of use. However, when I implemented this using Jotai, I had incorrect state problems, where state from other configs would show up. Ideally, I could use Jotai for this component like a tiny little global app where the state gets thrown away when it unmounts. I tried using a Provider from Jotai, but I wasn't sure where to place it in the hierarchy. Putting it above this component in the parent didn't work as that component didn't unmount and thus the state didn't reset.

I suspect I'm missing something, but I'm not sure what.

**Comment by dai-shi** [comment]
> I tried using a Provider from Jotai, but I wasn't sure where to place it in the hierarchy.

Hi, I'm curious why putting Jotai Provider at the same place as React Context Provider does not work.

I might be missing something in your use case, but in general jotai is designed to replace useState(useReducer)+useContext use cases. Maybe your use of `useQuery` complicates things?

**Reply by okbytes** [reply]
I don't think useQuery should complicate things unless the initial mount (no data) followed by data fetch messes Jotai up.

I had previously tried:

```jsx
function ConfigComponent(props) {
    const [state, setState] = useAtom(stateAtom)
    useQuery(GetDataDoc, {
        variables: { id: props.id },
        onCompleted: data => setState(data)
    })

    // ...state is used in the body of the function, too

    return (
        // ...lots of sub-components
    )
}

// ...further down in tree

function NestedComponent() {
    const [state, setState] = useAtom(stateAtom)

    // ...now this NestedComp can mutate the state for the whole config
    
    return (
        // ...markup
    )
}
```

However, I had trouble keeping the state fresh when `<ConfigComponent>` would receive different id values for loading a different configuration.

Is what you're saying to use my current structure and enlist Jotai to do the data drilling instead of context? Like this?

```jsx
function ConfigComponent(props) {
    const [state, dispatch] = useReducer(reducer, initialState);
    useQuery(GetDataDoc, {
        variables: { id: props.id },
        onCompleted: data => dispatch({ type: "replace", data })
    })

    //... state is used in the body of the function, too

    return (
        <Provider initialValues={{state, dispatch}}>
            {/*
                ...lots of sub-components
            */}
        </Provider>
    )
}
}

// ...further down in tree

function NestedComponent() {
    const [{state, dispatch}] = useAtom(stateAtom) // ?? not sure where atom lives

    return (
        // ...markup
    )
}
```

**Reply by dai-shi** [reply]
> Is what you're saying to use my current structure and enlist Jotai to do the data drilling instead of context?

No. I'm hoping Jotai Provider works in this case. You want to wrap you `ConfigComponent` in the first example (or any parent components that "remount") with Jotai's `Provider`.

**Reply by okbytes** [reply]
This appears to work exactly as I'd hoped (locally). I'm doing some more testing on a build environment to verify.

```jsx
function Parent() {
    // ...
    return (
        // ...markup
        <Provider key={id}>
            <ConfigComponent id={id} />
        </Provider>
    )
}

function ConfigComponent(props) {
    const [state, dispatch] = useReducerAtom(atom, reducer)
    useQuery(GetDataDoc, {
        variables: { id: props.id },
        onCompleted: data => dispatch({ type: "replace", data })
    })

    //... state is used in the body of the function, too

    return (
        // ...lots of sub-components
    )
}

// ...further down in tree

function NestedComponent() {
    const [state, dispatch] = useReducerAtom(atom, reducer)

    return (
        // ...markup
    )
}
```

**Reply by okbytes** [reply]
The above approach had a very nuanced bug. First, I had to scope the Provider in the Parent to avoid messing up global Jotai Atoms. Second, I had a nested query to refresh the data that would pull in state from the Scoped atom. It was returning previous state and causing a 422 error with our API (discovered by a customer).

I am not sure what the culprit is in this scenario. Maybe the Scoped Provider mixed with global un-Providered Atoms caused the staleness? I don't have time for a reproduction as I'm in the dog-house for this mistake being shipped. Noted here for future bug hunters.

**Reply by dai-shi** [reply]
Thanks for sharing your experience.

> I am not sure what the culprit is in this scenario.

Me neither. What I can say is the `scope` in Provider props and useAtom is pretty straightforward in terms of the implementation, so it's unlikely that `scope` itself is a technical issue. Can be a mental model issue, though.

> I had to scope the Provider in the Parent to avoid messing up global Jotai Atoms.

Yeah, this is understandable. Jotai's `scope` is very primitive and thus limited.
FYI, for your future investigation, there's a third-party library to overcome some of the limitations of jotai atoms: https://github.com/saasquatch/jotai-molecules
